### PR TITLE
Sklearn 1.6 support

### PR DESCRIPTION
**Pull request by Alexsandruss**
## Description

Changes:
- Set `SCIPY_ARRAY_API` to `1` which is required for sklearn>=1.6 if `array-api-compat` is installed in:
  - sklearnex config context
  - testing script
- Update input validation in GBT and AdaBoost daal4py estimators
- Update parameters validation logic for `pairwise_distances`
---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

N/A

### CODE REVIEW

File: .ci/pipeline/ci.yml [hunk 1]
```diff
@@ -134,6 +134,9 @@ jobs:
       Python3.13_Sklearn1.5:
         PYTHON_VERSION: '3.13'
         SKLEARN_VERSION: '1.5'
+      Python3.13_Sklearn1.6:
```
Comment (icfaust) [inline .ci/pipeline/ci.yml:137]: Please remove one of the other non-supported versions. Azure pipelines is already way overloaded with 3 hour runtimes.

File: daal4py/sklearn/ensemble/GBTDAAL.py [hunk 2]
```diff
@@ -128,9 +134,14 @@ def _check_params(self):
     def _more_tags(self):
         return {"allow_nan": self.allow_nan_}
 
+    def __sklearn_tags__(self):
```
Comment (icfaust) [inline daal4py/sklearn/ensemble/GBTDAAL.py:137]: I think this be behind a sklearn_check_version.

File: daal4py/sklearn/ensemble/GBTDAAL.py [hunk 3]
```diff
@@ -196,15 +207,18 @@ def fit(self, X, y):
...
-            X = check_array(X, dtype=[np.single, np.double], force_all_finite="allow-nan")
-
         # Check is fit had been called
         check_is_fitted(self, ["n_features_in_", "n_classes_"])
 
+        # Input validation
+        X = validate_data(
+            self,
+            X,
+            dtype=[np.single, np.double],
```
Comment (icfaust) [inline daal4py/sklearn/ensemble/GBTDAAL.py:217]:
```suggestion
            dtype=[np.float64, np.float32],
```
Order matters here for conversion, unless we always want to default to ```single```.  Also might as well follow convention elsewhere.

File: daal4py/sklearn/ensemble/GBTDAAL.py [hunk 4]
```diff
@@ -297,15 +311,18 @@ def fit(self, X, y):
...
-            X = check_array(X, dtype=[np.single, np.double], force_all_finite="allow-nan")
-
         # Check is fit had been called
         check_is_fitted(self, ["n_features_in_"])
 
+        # Input validation
+        X = validate_data(
+            self,
+            X,
+            dtype=[np.single, np.double],
```
Comment (icfaust) [inline daal4py/sklearn/ensemble/GBTDAAL.py:321]: Similar comment as before.

File: deselected_tests.yaml [hunk 5]
```diff
@@ -25,6 +25,20 @@
 #  will exclude deselection in versions 0.18.1, and 0.18.2 only.
 
 deselected_tests:
+  # sklearn 1.6 unsupported features
+  - linear_model/tests/test_base.py::test_linear_regression_sample_weight_consistency[42-True-None-X_shape1]
```
Comment (icfaust) [inline deselected_tests.yaml:29]: Some of these probably need tickets for follow-on analysis. Can be assigned to @Vika-F , @david-cortes-intel or @DDJHB since they have the most linreg/ridge experience.

**Review by icfaust — COMMENTED** [review]
Overall pretty good. @Alexsandruss are we going to do anything to control the sklearn version in conda-recipe to prevent CI failures, or was that bug a feature?

File: .ci/pipeline/ci.yml (hunk 1 above)
Comment (Alexsandruss) [reply .ci/pipeline/ci.yml:131]: I reduced matrix to last 5 versions of python and sklearn.

File: daal4py/sklearn/ensemble/GBTDAAL.py (hunk 2 above)
Comment (Alexsandruss) [reply daal4py/sklearn/ensemble/GBTDAAL.py]: Done.

File: daal4py/sklearn/ensemble/GBTDAAL.py (hunk 3 above)
Comment (Alexsandruss) [reply daal4py/sklearn/ensemble/GBTDAAL.py]: Fixed in all places for these estimators.

**Review by icfaust — APPROVED** [review]

**Comment by icfaust** [comment]
Merge on green CI

File: examples/sklearnex/knn_bf_regression_spmd.py [hunk 6]
```diff
@@ -80,6 +86,11 @@ def generate_X_y(par, coef_seed, data_seed):
 )
 print(
     "RMSE for entire rank {}: {}\n".format(
-        rank, mean_squared_error(y_test, dpt.to_numpy(y_predict), squared=False)
+        rank,
+        (
+            root_mean_squared_error(y_test, dpt.to_numpy(y_predict))
+            if sklearn_check_version("1.4")
+            else mean_squared_error(y_test, dpt.to_numpy(y_predict), squared=False)
+        ),
```
Comment (ethanglaser) [inline examples/sklearnex/knn_bf_regression_spmd.py:94]:
```suggestion
        rank, root_mean_squared_error(y_test, dpt.to_numpy(y_predict))
```

File: .ci/pipeline/ci.yml [hunk 7]
```diff
@@ -116,24 +116,21 @@ jobs:
   timeoutInMinutes: 120
   strategy:
     matrix:
-      Python3.9_Sklearn1.0:
```
Comment (ethanglaser) [inline .ci/pipeline/ci.yml:119]: In my opinion, it would make sense to continue checking sklearn 1.0 plus the latest versions - any thoughts on dropping/keeping this?

File: examples/sklearnex/knn_bf_regression_spmd.py [hunk 8]
```diff
@@ -21,7 +21,13 @@
...
 from mpi4py import MPI
 from numpy.testing import assert_allclose
-from sklearn.metrics import mean_squared_error
+
+from daal4py.sklearn._utils import sklearn_check_version
+
+if sklearn_check_version("1.4"):
+    from sklearn.metrics import root_mean_squared_error
+else:
+    from sklearn.metrics import mean_squared_error
```
[the remark this answers is not in the document]
Comment (Alexsandruss) [reply examples/sklearnex/knn_bf_regression_spmd.py]: RMSE was introduced in sklearn 1.4, so I moved to MSE without `squared`.

File: .ci/pipeline/ci.yml (hunk 7 above)
Comment (Alexsandruss) [reply .ci/pipeline/ci.yml:119]: Updated.

File: examples/sklearnex/knn_bf_regression_spmd.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ethanglaser) [reply examples/sklearnex/knn_bf_regression_spmd.py]: Good idea

**Comment by ethanglaser** [comment]
I guess this should also be updated: https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/doc/sources/quick-start.rst?plain=1#L377-L382

**Review by ethanglaser — APPROVED** [review]
LGTM. Is the SCIPY_ARRAY_API something we should eventually be supporting internally? Would a follow-up ticket make sense here?

**Comment by ethanglaser** [comment]
> Merge on green CI, we will need to also make tickets to upgrade private CI and update the public README.md about support.

I have added a ticket for private CI

File: deselected_tests.yaml (hunk 5 above)
Comment (Alexsandruss) [reply deselected_tests.yaml:29]: I created investigation ticket.

**Comment by Alexsandruss** [comment]
> I guess this should also be updated: https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/doc/sources/quick-start.rst?plain=1#L377-L382

Separate PR with update of supported versions and others docs would be better.
